File: ios/Halogen/Core/SwipePrefs.swift
import Foundation
import Observation
import SwiftUI

/// Configurable episode swipe actions — the full web vocabulary (ui-listview
/// `SwipeAction::ALL`), one per edge per page. Toggles resolve against the
/// row's live state (web `perform_episode_action`); on embedded accounts
/// device-download actions remap to server ones (the server IS this device).
enum SwipeAction: String, Codable, CaseIterable, Identifiable {
    case none
    case play
    case stream
    case togglePlayed = "toggle_played"
    case markPlayed = "mark_played"
    case markUnplayed = "mark_unplayed"
    case resetProgress = "reset_progress"
    case addToQueue = "add_to_queue"
    case removeFromQueue = "remove_from_queue"
    case toggleQueue = "toggle_queue"
    /// Remove from *this list's* playlist (queue or the viewed playlist).
    case removeFromList = "remove_from_list"
    case addToPlaylist = "add_to_playlist"
    /// Keeps the pre-parity "download" token so saved prefs still decode.
    case downloadToDevice = "download"
    case redownloadDevice = "redownload_device"
    case removeDownload = "remove_download"
    case toggleDownload = "toggle_download"
    case downloadOnServer = "download_on_server"
    case removeFromServer = "remove_from_server"
    case toggleServerDownload = "toggle_server_download"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .none: return "None"
        case .play: return "Play"
        case .stream: return "Stream"
        case .togglePlayed: return "Toggle played"
        case .markPlayed: return "Mark played"
        case .markUnplayed: return "Mark unplayed"
        case .resetProgress: return "Reset progress"
        case .addToQueue: return "Add to queue"
        case .removeFromQueue: return "Remove from queue"
        case .toggleQueue: return "Toggle queue"
        case .removeFromList: return "Remove from list"
        case .addToPlaylist: return "Add to playlist"
        case .downloadToDevice: return "Download"
        case .redownloadDevice: return "Re-download"
        case .removeDownload: return "Remove download"
        case .toggleDownload: return "Toggle download"
        case .downloadOnServer: return "Download on server"
        case .removeFromServer: return "Remove from server"
        case .toggleServerDownload: return "Toggle server download"
        }
    }

    var systemImage: String {
        switch self {
        case .none: return "slash.circle"
        case .play: return "play.fill"
        case .stream: return "dot.radiowaves.left.and.right"
        case .togglePlayed: return "checkmark.circle.badge.questionmark"
        case .markPlayed: return "checkmark.circle"
        case .markUnplayed: return "circle"
        case .resetProgress: return "arrow.counterclockwise"
        case .addToQueue: return "text.badge.plus"
        case .removeFromQueue: return "text.badge.minus"
        case .toggleQueue: return "text.badge.checkmark"
        case .removeFromList: return "minus.circle"
        case .addToPlaylist: return "music.note.list"
        case .downloadToDevice: return "arrow.down.to.line.circle"
        case .redownloadDevice: return "arrow.triangle.2.circlepath"
        case .removeDownload: return "iphone.slash"
        case .toggleDownload: return "arrow.down.circle.dotted"
        case .downloadOnServer: return "icloud.and.arrow.down"
        case .removeFromServer: return "icloud.slash"
        case .toggleServerDownload: return "icloud.circle"
        }
    }

    /// Destructive/removal actions get the red tint; additive ones blue-ish.
    var tint: Color {
        switch self {
        case .none: return .gray
        case .play, .stream: return .indigo
        case .togglePlayed, .markPlayed, .markUnplayed, .resetProgress: return .green
        case .addToQueue, .toggleQueue: return .blue
        case .removeFromQueue, .removeFromList: return .red
        case .addToPlaylist: return .purple
        case .downloadToDevice, .redownloadDevice, .toggleDownload,
            .downloadOnServer, .toggleServerDownload:
            return .orange
        case .removeDownload, .removeFromServer: return .red
        }
    }
}

/// The pages whose swipes are configurable — all six of the web's
/// (webui/config swipe.rs `SwipePage::ALL`).
enum SwipePage: String, Codable, CaseIterable, Identifiable {
    case latest
    case queue
    case podcastEpisodes = "podcast_episodes"
    case downloads
    case history
    case playlist

    var id: String { rawValue }

    var label: String {
        switch self {
        case .latest: return "Latest"
        case .queue: return "Queue"
        case .podcastEpisodes: return "Podcast episodes"
        case .downloads: return "Downloads"
        case .history: return "History"
        case .playlist: return "Playlist"
        }
    }

    /// Actions offered for this page, in menu order — `removeFromList` only
    /// where the list IS a playlist (web: `SwipePage::allowed_actions`).
    var allowedActions: [SwipeAction] {
        SwipeAction.allCases.filter { action in
            action != .removeFromList || self == .queue || self == .playlist
        }
    }
}

struct SwipePrefs: Codable, Equatable {
    struct PagePrefs: Codable, Equatable {
        var leading: SwipeAction
        var trailing: SwipeAction
    }

    var pages: [String: PagePrefs]

    /// The web defaults (webui/config swipe.rs): `leading` here is the
    /// web's `left` (fires on a swipe-RIGHT gesture), `trailing` its `right`.
    static let `default` = SwipePrefs(
        pages: [
            SwipePage.latest.rawValue: PagePrefs(
                leading: .addToQueue, trailing: .downloadToDevice),
            SwipePage.queue.rawValue: PagePrefs(
                leading: .removeFromQueue, trailing: .downloadToDevice),
            SwipePage.podcastEpisodes.rawValue: PagePrefs(
                leading: .addToPlaylist, trailing: .downloadToDevice),
            SwipePage.downloads.rawValue: PagePrefs(
                leading: .addToPlaylist, trailing: .redownloadDevice),
            SwipePage.history.rawValue: PagePrefs(
                leading: .addToPlaylist, trailing: .downloadToDevice),
            SwipePage.playlist.rawValue: PagePrefs(
                leading: .removeFromList, trailing: .downloadToDevice),
        ]
    )

    /// Unset pages (saved prefs predating a page) fall back to the web
    /// default, not to a dead swipe.
    func page(_ page: SwipePage) -> PagePrefs {
        pages[page.rawValue]
            ?? Self.default.pages[page.rawValue]
            ?? PagePrefs(leading: .none, trailing: .none)
    }
}

/// Reactive holder + per-account persistence.
@MainActor
@Observable
final class SwipePrefsModel {
    private static let key = "swipe-prefs"

    private(set) var prefs: SwipePrefs = .default
    private let store: LocalStore?

    init(store: LocalStore?) {
        self.store = store
    }

    func load() async {
        if let store, let saved = await store.load(SwipePrefs.self, key: Self.key) {
            prefs = saved
        }
    }

    func update(_ page: SwipePage, leading: SwipeAction, trailing: SwipeAction) {
        prefs.pages[page.rawValue] = SwipePrefs.PagePrefs(leading: leading, trailing: trailing)
        let snapshot = prefs
        Task { [store] in await store?.save(snapshot, key: Self.key) }
    }
}

extension View {
    /// Attach the configured leading/trailing swipe actions for `page`.
    /// `context` resolves the list-scoped actions (remove-from-list) the same
    /// way the row menu does.
    func configuredSwipes(
        _ page: SwipePage, episode: EpisodeData, core: HalogenCore,
        context: EpisodeMenuContext = .browse
    ) -> some View {
        let prefs =
            core.models?.swipes.prefs.page(page)
            ?? SwipePrefs.PagePrefs(leading: .none, trailing: .none)
        return
            self
            .swipeActions(edge: .leading, allowsFullSwipe: true) {
                SwipeActionButton(
                    action: prefs.leading, episode: episode, core: core, context: context)
            }
            .swipeActions(edge: .trailing, allowsFullSwipe: false) {
                SwipeActionButton(
                    action: prefs.trailing, episode: episode, core: core, context: context)
            }
    }
}

private struct SwipeActionButton: View {
    let action: SwipeAction
    let episode: EpisodeData
    let core: HalogenCore
    var context: EpisodeMenuContext = .browse

    var body: some View {
        if action != .none {
            Button {
                perform()
            } label: {
                Label(action.label, systemImage: action.systemImage)
            }
            .tint(action.tint)
        }
    }

    /// The action→primitive mapping — one place, mirroring the web's
    /// `perform_episode_action` (webui/episode-list action.rs), embedded
    /// device→server remaps included.
    private func perform() {
        guard let models = core.models else { return }
        let embedded = core.isEmbeddedAccount
        switch action {
        case .none:
            break
        case .play:
            models.player.play(episode, context: context.playbackContext)
        case .stream:
            models.player.stream(episode, context: context.playbackContext)
        case .togglePlayed:
            let status = models.playbacks.status(for: episode)
            models.playbacks.markPlayed(episode, played: status != .finished)
        case .markPlayed:
            models.playbacks.markPlayed(episode, played: true)
        case .markUnplayed:
            models.playbacks.markPlayed(episode, played: false)
        case .resetProgress:
            models.playbacks.setCursor(episode, cursor: 0)
        case .addToQueue:
            models.queue.add(episode)
        case .removeFromQueue:
            models.queue.remove(episode)
        case .toggleQueue:
            if models.queue.contains(episode) {
                models.queue.remove(episode)
            } else {
                models.queue.add(episode)
            }
        case .removeFromList:
            switch context {
            case .queue:
                models.queue.remove(episode)
            case .playlist(_, let model):
                model.remove(episode)
            case .browse:
                break  // not offered on browse pages (allowedActions)
            }
        case .addToPlaylist:
            models.playlists.requestPick([episode])
        case .downloadToDevice:
            if embedded {
                models.serverDownloads.download(episode)
            } else {
                models.device.download(episode)
            }
        case .redownloadDevice:
            if embedded {
                redownloadServer()
            } else if !core.isOffline {
                // Web rule: never destroy a local copy you can't re-pull.
                models.device.remove(episode.id)
                models.device.download(episode)
            } else {
                // The refusal is right; the silence wasn't — the swipe just
                // animated closed with no effect.
                ToastCenter.shared.error("You're offline — re-download needs a connection.")
            }
        case .removeDownload:
            if embedded {
                removeServer()
            } else {
                models.device.remove(episode.id)
            }
        case .toggleDownload:
            if embedded {
                toggleServer()
            } else if models.device.state(of: episode.id) == .downloaded {
                models.device.remove(episode.id)
            } else {
                models.device.download(episode)
            }
        case .downloadOnServer:
            models.serverDownloads.download(episode)
        case .removeFromServer:
            removeServer()
        case .toggleServerDownload:
            toggleServer()
        }
    }

    private func removeServer() {
        core.enqueueMutation(.removeServerDownload(episodeId: episode.id)) {
            core.models?.serverDownloads.markRemovedLocally(episode.id)
        }
    }

    /// Remove-then-trigger, in outbox order (web: RedownloadOnServer).
    private func redownloadServer() {
        Task {
            guard
                await core.ensureQueuedBatch([
                    .removeServerDownload(episodeId: episode.id),
                    .triggerDownload(episodeId: episode.id),
                ])
            else { return }
            core.models?.serverDownloads.watch(episode.id)
        }
    }

    private func toggleServer() {
        if core.models?.serverDownloads.isDownloaded(episode) == true {
            removeServer()
        } else {
            core.models?.serverDownloads.download(episode)
        }
    }
}

File: android/app/src/main/kotlin/org/fgsec/halogen/core/SwipePrefs.kt
package org.fgsec.halogen.core

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.fgsec.halogen.components.EpisodeMenuContext
import org.fgsec.halogen.components.ToastCenter
import org.fgsec.halogen.storage.LocalStore
import org.fgsec.halogen.storage.OutboxOp
import org.fgsec.halogen.wire.EpisodeData
import org.fgsec.halogen.wire.PlaybackStatus

/// Configurable episode swipe actions — the full web vocabulary (webui/listview
/// `SwipeAction::ALL`), one per edge per page. Toggles resolve against the row's live
/// state; embedded accounts remap device-download actions to their server counterparts.
@Serializable
enum class SwipeAction(val token: String) {
    @SerialName("none") None("none"),
    @SerialName("play") Play("play"),
    @SerialName("stream") Stream("stream"),
    @SerialName("toggle_played") TogglePlayed("toggle_played"),
    @SerialName("mark_played") MarkPlayed("mark_played"),
    @SerialName("mark_unplayed") MarkUnplayed("mark_unplayed"),
    @SerialName("reset_progress") ResetProgress("reset_progress"),
    @SerialName("add_to_queue") AddToQueue("add_to_queue"),
    @SerialName("remove_from_queue") RemoveFromQueue("remove_from_queue"),
    @SerialName("toggle_queue") ToggleQueue("toggle_queue"),
    /// Remove from *this list's* playlist (queue or the viewed playlist).
    @SerialName("remove_from_list") RemoveFromList("remove_from_list"),
    @SerialName("add_to_playlist") AddToPlaylist("add_to_playlist"),
    /// Keeps the pre-parity "download" token so saved prefs still decode.
    @SerialName("download") DownloadToDevice("download"),
    @SerialName("redownload_device") RedownloadDevice("redownload_device"),
    @SerialName("remove_download") RemoveDownload("remove_download"),
    @SerialName("toggle_download") ToggleDownload("toggle_download"),
    @SerialName("download_on_server") DownloadOnServer("download_on_server"),
    @SerialName("remove_from_server") RemoveFromServer("remove_from_server"),
    @SerialName("toggle_server_download") ToggleServerDownload("toggle_server_download");

    val label: String
        get() = when (this) {
            None -> "None"
            Play -> "Play"
            Stream -> "Stream"
            TogglePlayed -> "Toggle played"
            MarkPlayed -> "Mark played"
            MarkUnplayed -> "Mark unplayed"
            ResetProgress -> "Reset progress"
            AddToQueue -> "Add to queue"
            RemoveFromQueue -> "Remove from queue"
            ToggleQueue -> "Toggle queue"
            RemoveFromList -> "Remove from list"
            AddToPlaylist -> "Add to playlist"
            DownloadToDevice -> "Download"
            RedownloadDevice -> "Re-download"
            RemoveDownload -> "Remove download"
            ToggleDownload -> "Toggle download"
            DownloadOnServer -> "Download on server"
            RemoveFromServer -> "Remove from server"
            ToggleServerDownload -> "Toggle server download"
        }

    /// SF Symbol name — components/Icons.kt maps it to a Material icon.
    val systemImage: String
        get() = when (this) {
            None -> "slash.circle"
            Play -> "play.fill"
            Stream -> "dot.radiowaves.left.and.right"
            TogglePlayed -> "checkmark.circle.badge.questionmark"
            MarkPlayed -> "checkmark.circle"
            MarkUnplayed -> "circle"
            ResetProgress -> "arrow.counterclockwise"
            AddToQueue -> "text.badge.plus"
            RemoveFromQueue -> "text.badge.minus"
            ToggleQueue -> "text.badge.checkmark"
            RemoveFromList -> "minus.circle"
            AddToPlaylist -> "music.note.list"
            DownloadToDevice -> "arrow.down.to.line.circle"
            RedownloadDevice -> "arrow.triangle.2.circlepath"
            RemoveDownload -> "iphone.slash"
            ToggleDownload -> "arrow.down.circle.dotted"
            DownloadOnServer -> "icloud.and.arrow.down"
            RemoveFromServer -> "icloud.slash"
            ToggleServerDownload -> "icloud.circle"
        }

    /// Destructive/removal actions get the red tint; additive ones blue-ish.
    /// Semantic name only — components/Theme.kt maps it to a Color.
    val tint: SwipeTint
        get() = when (this) {
            None -> SwipeTint.Gray
            Play, Stream -> SwipeTint.Indigo
            TogglePlayed, MarkPlayed, MarkUnplayed, ResetProgress -> SwipeTint.Green
            AddToQueue, ToggleQueue -> SwipeTint.Blue
            RemoveFromQueue, RemoveFromList -> SwipeTint.Red
            AddToPlaylist -> SwipeTint.Purple
            DownloadToDevice, RedownloadDevice, ToggleDownload,
            DownloadOnServer, ToggleServerDownload -> SwipeTint.Orange
            RemoveDownload, RemoveFromServer -> SwipeTint.Red
        }
}

/// The iOS system-color vocabulary the swipe tints use.
enum class SwipeTint { Gray, Indigo, Green, Blue, Red, Purple, Orange }

/// The pages whose swipes are configurable — all six of the web's
/// (webui/config swipe.rs `SwipePage::ALL`).
@Serializable
enum class SwipePage(val token: String) {
    @SerialName("latest") Latest("latest"),
    @SerialName("queue") Queue("queue"),
    @SerialName("podcast_episodes") PodcastEpisodes("podcast_episodes"),
    @SerialName("downloads") Downloads("downloads"),
    @SerialName("history") History("history"),
    @SerialName("playlist") Playlist("playlist");

    val label: String
        get() = when (this) {
            Latest -> "Latest"
            Queue -> "Queue"
            PodcastEpisodes -> "Podcast episodes"
            Downloads -> "Downloads"
            History -> "History"
            Playlist -> "Playlist"
        }

    /// Actions offered for this page, in menu order — `RemoveFromList` only
    /// where the list IS a playlist (web: `SwipePage::allowed_actions`).
    val allowedActions: List<SwipeAction>
        get() = SwipeAction.entries.filter { action ->
            action != SwipeAction.RemoveFromList || this == Queue || this == Playlist
        }
}

@Serializable
data class SwipePrefs(val pages: Map<String, PagePrefs>) {
    @Serializable
    data class PagePrefs(val leading: SwipeAction, val trailing: SwipeAction)

    /// Unset pages (saved prefs predating a page) fall back to the web
    /// default, not to a dead swipe.
    fun page(page: SwipePage): PagePrefs =
        pages[page.token]
            ?: default.pages[page.token]
            ?: PagePrefs(SwipeAction.None, SwipeAction.None)

    companion object {
        /// The web defaults (webui/config swipe.rs): `leading` here is the
        /// web's `left` (fires on a swipe-RIGHT gesture), `trailing` its `right`.
        val default = SwipePrefs(
            pages = mapOf(
                SwipePage.Latest.token to
                    PagePrefs(SwipeAction.AddToQueue, SwipeAction.DownloadToDevice),
                SwipePage.Queue.token to
                    PagePrefs(SwipeAction.RemoveFromQueue, SwipeAction.DownloadToDevice),
                SwipePage.PodcastEpisodes.token to
                    PagePrefs(SwipeAction.AddToPlaylist, SwipeAction.DownloadToDevice),
                SwipePage.Downloads.token to
                    PagePrefs(SwipeAction.AddToPlaylist, SwipeAction.RedownloadDevice),
                SwipePage.History.token to
                    PagePrefs(SwipeAction.AddToPlaylist, SwipeAction.DownloadToDevice),
                SwipePage.Playlist.token to
                    PagePrefs(SwipeAction.RemoveFromList, SwipeAction.DownloadToDevice),
            )
        )
    }
}

/// Reactive holder + per-account persistence.
class SwipePrefsModel(
    private val store: LocalStore?,
    private val scope: CoroutineScope = MainScope(),
) {
    var prefs: SwipePrefs by mutableStateOf(SwipePrefs.default)
        private set

    suspend fun load() {
        val saved = store?.load<SwipePrefs>(KEY) ?: return
        prefs = saved
    }

    fun update(page: SwipePage, leading: SwipeAction, trailing: SwipeAction) {
        prefs = prefs.copy(
            pages = prefs.pages + (page.token to SwipePrefs.PagePrefs(leading, trailing)))
        val snapshot = prefs
        scope.launch { store?.save(snapshot, KEY) }
    }

    private companion object {
        const val KEY = "swipe-prefs"
    }
}

/// The action→primitive mapping — one place, mirroring the web's
/// `perform_episode_action` (webui/episode-list action.rs), embedded
/// device→server remaps included. The Compose swipe container
/// (components/SwipeActions.kt) and row menus both call through here.
fun performSwipeAction(
    action: SwipeAction,
    episode: EpisodeData,
    core: HalogenCore,
    context: EpisodeMenuContext = EpisodeMenuContext.Browse,
) {
    val models = core.models ?: return
    val embedded = core.isEmbeddedAccount
    when (action) {
        SwipeAction.None -> {}
        SwipeAction.Play ->
            models.player.play(episode, context.playbackContext)
        SwipeAction.Stream ->
            models.player.stream(episode, context.playbackContext)
        SwipeAction.TogglePlayed -> {
            val status = models.playbacks.status(episode)
            models.playbacks.markPlayed(episode, status != PlaybackStatus.Finished)
        }
        SwipeAction.MarkPlayed -> models.playbacks.markPlayed(episode, true)
        SwipeAction.MarkUnplayed -> models.playbacks.markPlayed(episode, false)
        SwipeAction.ResetProgress -> models.playbacks.setCursor(episode, 0uL)
        SwipeAction.AddToQueue -> models.queue.add(episode)
        SwipeAction.RemoveFromQueue -> models.queue.remove(episode)
        SwipeAction.ToggleQueue ->
            if (models.queue.contains(episode)) models.queue.remove(episode)
            else models.queue.add(episode)
        SwipeAction.RemoveFromList -> when (context) {
            is EpisodeMenuContext.Queue -> models.queue.remove(episode)
            is EpisodeMenuContext.Playlist -> context.model.remove(episode)
            is EpisodeMenuContext.Browse -> {} // not offered on browse pages (allowedActions)
        }
        SwipeAction.AddToPlaylist -> models.playlists.requestPick(listOf(episode))
        SwipeAction.DownloadToDevice ->
            if (embedded) models.serverDownloads.download(episode)
            else models.device.download(episode)
        SwipeAction.RedownloadDevice -> when {
            embedded -> redownloadServer(core, episode)
            !core.isOffline -> {
                // Web rule: never destroy a local copy you can't re-pull.
                models.device.remove(episode.id)
                models.device.download(episode)
            }
            // The refusal is right; the silence wasn't — the swipe just
            // animated closed with no effect.
            else -> ToastCenter.error("You're offline — re-download needs a connection.")
        }
        SwipeAction.RemoveDownload ->
            if (embedded) removeServer(core, episode)
            else models.device.remove(episode.id)
        SwipeAction.ToggleDownload -> when {
            embedded -> toggleServer(core, episode)
            models.device.stateOf(episode.id) == DeviceDownloads.State.Downloaded ->
                models.device.remove(episode.id)
            else -> models.device.download(episode)
        }
        SwipeAction.DownloadOnServer -> models.serverDownloads.download(episode)
        SwipeAction.RemoveFromServer -> removeServer(core, episode)
        SwipeAction.ToggleServerDownload -> toggleServer(core, episode)
    }
}

private fun removeServer(core: HalogenCore, episode: EpisodeData) {
    core.enqueueMutation(OutboxOp.Kind.RemoveServerDownload(episode.id)) {
        core.models?.serverDownloads?.markRemovedLocally(episode.id)
    }
}

/// Remove-then-trigger, in outbox order (web: RedownloadOnServer).
private fun redownloadServer(core: HalogenCore, episode: EpisodeData) {
    core.scope.launch {
        if (!core.ensureQueuedBatch(listOf(OutboxOp.Kind.RemoveServerDownload(episode.id), OutboxOp.Kind.TriggerDownload(episode.id)))) return@launch
        core.models?.serverDownloads?.watch(episode.id)
    }
}

private fun toggleServer(core: HalogenCore, episode: EpisodeData) {
    if (core.models?.serverDownloads?.isDownloaded(episode) == true) {
        removeServer(core, episode)
    } else {
        core.models?.serverDownloads?.download(episode)
    }
}
